### PR TITLE
EdgeHub: Add code to drop connections after the tokens expire

### PR DESCRIPTION
**Pull request by varunpuranik**
- Add a Credentials cache to store the latest credentials from the connection
- Add a thread that authenticates those credentials every 5 mins (same as IoThub). If the credentials were authenticated by connecting to IoTHub, then just check if the token is expired.
- Rename CredentialsStore to CredentialsCache (since it will be in memory in most cases)

### CODE REVIEW

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs [hunk 1]
```diff
@@ -77,17 +87,10 @@ await device.CloudConnection.Filter(cp => cp.IsActive)
...
-            return this.devices.TryGetValue(Preconditions.CheckNonWhiteSpace(id, nameof(id)), out ConnectedDevice device)
-                ? device.DeviceConnection.Filter(dp => dp.IsActive).Map(dc => dc.ClientCredentials)
-                : Option.None<IClientCredentials>();
-        }
-
         public Option<IDeviceProxy> GetDeviceConnection(string id)
         {
             return this.devices.TryGetValue(Preconditions.CheckNonWhiteSpace(id, nameof(id)), out ConnectedDevice device)
-                ? device.DeviceConnection.Filter(dp => dp.IsActive).FlatMap(d => d.DeviceProxy)
+                ? device.DeviceConnection.Filter(dp => dp.IsActive).Map(d => d.DeviceProxy)
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionManager.cs]: Is `d.DeviceProxy` no longer an `Option`?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionReauthenticator.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                }
+                else
+                {
+                    Log.LogWarning((int)EventIds.ClientCredentialsResult, $"Reauthenticating client {identity.Id} failed, removing client connection");
+                }
+            }
+
+            public static void ClientCredentialsNotFound(IIdentity identity)
+            {
+                Log.LogError((int)EventIds.ClientCredentialsResult, $"Unable to reauthenticate client {identity.Id} as the client credentials were not found, removing client connection");
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionReauthenticator.cs]: Is this really an error or a warning? It's fine if it's actually an error, just asking?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/CredentialsCache.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+
+    public class CredentialsCache : ICredentialsCache
+    {
+        readonly IDictionary<string, IClientCredentials> cache = new ConcurrentDictionary<string, IClientCredentials>();
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/CredentialsCache.cs]: Does this need to be a ConcurrentDictionary, if it's always guarded by a lock?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs [hunk 4]
```diff
@@ -118,6 +118,13 @@ public static async Task<int> MainAsync(IConfigurationRoot configuration)
             ConfigUpdater configUpdater = await container.Resolve<Task<ConfigUpdater>>();
             await configUpdater.Init(configSource);
 
+            if (!Enum.TryParse(configuration.GetValue("AuthenticationMode", string.Empty), true, out AuthenticationMode authenticationMode)
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs]: Should this be case insensitive?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionReauthenticator.cs (hunk 2 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ConnectionReauthenticator.cs]: Good catch, should be a warning.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs (hunk 4 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs]: It is, the second parameter "true" is for ignoreCase

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/CredentialsCache.cs (hunk 3 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/CredentialsCache.cs]: ah, I wanted to revert that.. (and thought I did). Will fix.

**Comment by varunpuranik** [comment]
@myagley - I addressed your comments. ptal.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeTokenAuthenticator.cs [hunk 5]
```diff
@@ -19,64 +19,40 @@ public class DeviceScopeTokenAuthenticator : IAuthenticator
...
+                return await serviceIdentity.Map(s => this.AuthenticateInternalAsync(tokenCredentials, s))
+                    .GetOrElse(() => this.underlyingAuthenticator.AuthenticateAsync(clientCredentials));
             }
-            catch (Exception ex)
+            catch (Exception e)
             {
-                Events.ErrorRemovingConnection(ex, id);
+                Events.ErrorAuthenticating(e, clientCredentials);
+                return await this.underlyingAuthenticator.AuthenticateAsync(clientCredentials);
             }
```
Comment (ancaantochi) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeTokenAuthenticator.cs]: Can this be called only when AuthenticateInternalAsync fails, no need to call it if the underlyingAuthenticator failed?

Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeTokenAuthenticator.cs]: Good point, will fix.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs (hunk 4 above)
Comment (myagley) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs]: Ah I see.

**Pull request merged**